File: packages/reference/android/src/test/java/github/penguin/reference/InstanceConverterTest.java
package github.penguin.reference;

import org.junit.Before;
import org.junit.Test;

import java.util.Collections;

import github.penguin.reference.TestClasses.TestMessenger;
import github.penguin.reference.reference.InstanceConverter.StandardInstanceConverter;
import github.penguin.reference.reference.PairedInstance;

import static org.junit.Assert.assertEquals;

public class InstanceConverterTest {
  private static final StandardInstanceConverter converter = new StandardInstanceConverter();
  private static TestMessenger testMessenger;

  @Before
  public void setUp() {
    testMessenger = new TestMessenger();
  }

  @Test
  public void convertInstancesToPairedInstances_handlesPairedObject()
      throws Exception {
    testMessenger.onReceiveCreateNewInstancePair("test_channel",
        new PairedInstance("test_id"),
        Collections.emptyList(), true);

    assertEquals(new PairedInstance("test_id"),
        converter.convertInstancesToPairedInstances(testMessenger, testMessenger.testHandler.testClassInstance)
    );
  }

  @Test
  public void convertInstancesToPairedInstances_handlesUnpairedObject() {
    assertEquals("potato", converter.convertInstancesToPairedInstances(testMessenger, "potato"));
  }

  @Test
  public void convertPairedInstancesToInstances_handlesPairedInstance() throws Exception {
    testMessenger.onReceiveCreateNewInstancePair(
        "test_channel",
        new PairedInstance("test_id"),
        Collections.emptyList(), true
    );

    assertEquals(
        converter.convertPairedInstancesToInstances(testMessenger, new PairedInstance("test_id")),
        testMessenger.testHandler.testClassInstance
    );
  }

  @Test
  public void convertPairedInstancesToInstances_handlesNewUnpairedObject() {
    assertEquals("potato", converter.convertPairedInstancesToInstances(testMessenger, "potato"));
  }
}

File: packages/reference/test/instance_converter_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:reference/reference.dart';

import 'test_classes.dart';

void main() {
  group('$StandardInstanceConverter', () {
    const StandardInstanceConverter converter = StandardInstanceConverter();
    late TestMessenger testMessenger;

    setUp(() {
      testMessenger = TestMessenger();
    });

    test('convertInstances handles paired Object', () {
      testMessenger.onReceiveCreateNewInstancePair(
        'test_channel',
        const PairedInstance('test_id'),
        <Object>[],
        owner: true,
      );

      expect(
        converter.convertInstances(
          testMessenger.instanceManager,
          testMessenger.testHandler.testClassInstance,
        ),
        const PairedInstance('test_id'),
      );
    });

    test('convertInstances handles unpaired object', () {
      expect(
        converter.convertInstances(testMessenger.instanceManager, 'potato'),
        equals('potato'),
      );
    });

    test('convertPairedInstances handles $PairedInstance', () {
      const PairedInstance pairedInstance = PairedInstance('test_id');
      testMessenger.onReceiveCreateNewInstancePair(
        'test_channel',
        pairedInstance,
        <Object>[],
        owner: true,
      );

      expect(
        converter.convertPairedInstances(
          testMessenger.instanceManager,
          pairedInstance,
        ),
        testMessenger.testHandler.testClassInstance,
      );
    });

    test('convertPairedInstances handles unpaired object', () async {
      expect(
        converter.convertPairedInstances(
          testMessenger.instanceManager,
          'apple',
        ),
        equals('apple'),
      );
    });
  });
}
